File: python/백준/백트래킹/15663 N과 M (9).py
def back_tracking(arr, now_length, num_arr, used_index_arr, result):
    if now_length == M:
        result.append(arr[:])  # 결과 리스트에 현재 배열 복사본 저장
        return

    prev = -1  # 이전에 선택한 숫자를 추적
    for idx in range(len(num_arr)):
        if not used_index_arr[idx] and num_arr[idx] != prev:
            used_index_arr[idx] = True
            arr.append(num_arr[idx])
            back_tracking(arr, now_length + 1, num_arr, used_index_arr, result)
            used_index_arr[idx] = False
            arr.pop()
            prev = num_arr[idx]  # 현재 숫자를 이전 값으로 업데이트

N, M = map(int, input().split())
num_arr = list(map(int, input().split()))
used_index_arr = [False] * len(num_arr)
result = []

back_tracking([], 0, sorted(num_arr), used_index_arr, result)

for r in result:
    print(" ".join(map(str, r)))

File: cpp/백준/백트래킹/15663 N과 M (9).cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N, M;
vector<int> num_arr;
vector<bool> used_index_arr;
vector<int> arr;

void back_tracking(int now_length) {
    if (now_length == M) {
        for (int i = 0; i < M; ++i) {
            cout << arr[i];
            if (i + 1 < M) cout << ' ';
        }
        cout << '\n';
        return;
    }

    int prev = -1;
    for (int idx = 0; idx < (int)num_arr.size(); ++idx) {
        if (!used_index_arr[idx] && num_arr[idx] != prev) {
            used_index_arr[idx] = true;
            arr.push_back(num_arr[idx]);
            back_tracking(now_length + 1);
            used_index_arr[idx] = false;
            arr.pop_back();
            prev = num_arr[idx];
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;
    num_arr.resize(N);
    for (int i = 0; i < N; ++i) cin >> num_arr[i];
    sort(num_arr.begin(), num_arr.end());
    used_index_arr.assign(N, false);

    back_tracking(0);
    return 0;
}
